zerver/lib/api_parameter_parsing.py

```diff
@@ -0,0 +1,392 @@
+    possible not leaking side effects to the global state. Side effects should
+    be executed separately after the ViewFuncInfo is returned.
+    """
+    type_hints = get_type_hints(view_func, include_extras=True)
+    parameters = inspect.signature(view_func).parameters
+    view_func_full_name = f"{view_func.__module__}.{view_func.__name__}"
+
+    process_parameters: List[FuncParam[object]] = []
+
+    for param_name, parameter in parameters.items():
+        assert param_name in type_hints
+        if param_name in IGNORED_PARAMS:
+            continue
+        param_info = parse_single_parameter(
+            param_name=param_name, param_type=type_hints[param_name], parameter=parameter
+        )
+        process_parameters.append(param_info)
+
+    return ViewFuncInfo(
+        view_func_full_name=view_func_full_name,
+        parameters=process_parameters,
+        no_nesting=True,
+    )
+
+
+ERROR_TEMPLATES = {
+    "int_type": _("{var_name} is not an integer"),
+    "int_parsing": _("{var_name} is not an integer"),
+    "string_type": _("{var_name} is not a string"),
+    "datetime_type": _("{var_name} is not a date"),
+    "bool_type": _("{var_name} is not a boolean"),
+    "list_type": _("{var_name} is not a list"),
+    "dict_type": _("{var_name} is not a dict"),
+    "float_type": _("{var_name} is not a float"),
+    "string_too_long": _("{var_name} is too long (limit: {max_length} characters)"),
+    "string_too_short": _("{item} cannot be blank."),
+    "less_than": _("{var_name} is too small"),
+    "greater_than": _("{var_name} is too large"),
+    "json_type": _("{var_name} is not valid JSON"),
+    "json_invalid": _("{var_name} is not valid JSON"),
+    "unexpected_keyword_argument": _('Argument "{argument}" at {var_name} is unexpected'),
+}
+
+MISSING_ERROR_TEMPLATE = """
+Pydantic validation error of type "{error_type}" does not have the corresponding
+error message template. We expect that every validation error is formatted into
+a client-facing error message. Consider adding this type to ERROR_TEMPLATES with
+the appropriate internationalized error message.
+Documentation for "{error_type}" can be found at {url}.
+"""
+
+
+def get_func_param(info: ValidationInfo) -> FuncParam[object]:
+    assert (
+        info.context is not None and "func_param" in info.context
+    ), "The FuncParam object should always be passed as context when validating"
+    return info.context["func_param"]
+
+
+def parse_param_to_type(parameter: FuncParam[T], value: object) -> T:
+    try:
+        return TypeAdapter(parameter.param_type).validate_python(
+            value, strict=parameter.strict_mode, context={"func_param": parameter}
+        )
+    except ValidationError as exc:
+        for error in exc.errors():
+            error_template = ERROR_TEMPLATES.get(error["type"])
+            var_name = parameter.request_var_name + "".join(
+                [f'["{loc}"]' if isinstance(loc, str) else f"[{loc}]" for loc in error["loc"]]
```

Comment:
Just `f"[{loc!r}]"` would handle numbers and strings with correct escaping.

Reply:
Right, thanks!